### PR TITLE
Groth16test

### PR DESCRIPTION
**Pull request by joyqvq**
## Description 

open for early review. see some open questions in comments
## Test Plan 

How did you test the new or updated feature?

---
If your changes are not user-facing and not a breaking change, you can skip the following section. Otherwise, please indicate what changed, and then add to the Release Notes section as highlighted during the release process.

### Type of Change (Check all that apply)

- [ ] protocol change
- [ ] user-visible impact
- [ ] breaking change for a client SDKs
- [ ] breaking change for FNs (FN binary must upgrade)
- [ ] breaking change for validators or node operators (must upgrade binaries)
- [ ] breaking change for on-chain data layout
- [ ] necessitate either a data wipe or data migration

### CODE REVIEW

File: crates/sui-core/src/consensus_validator.rs [hunk 1]
```diff
@@ -89,11 +92,12 @@ impl TransactionValidator for SuiTxValidator {
         let cert_count = cert_batch.len();
         let ckpt_count = ckpt_batch.len();
         let epoch_store = self.epoch_store.clone();
+        let jwk = self.google_jwk_as_bytes.clone();
         Handle::current()
             .spawn_blocking(move || {
                 epoch_store
                     .signature_verifier
-                    .verify_certs_and_checkpoints(cert_batch, ckpt_batch)
+                    .verify_certs_and_checkpoints(cert_batch, ckpt_batch, jwk)
```
Comment (mystenmark) [inline crates/sui-core/src/consensus_validator.rs:100]: I think we could live with non-deterministic signing due to each validator having a different view of the JWK, at least while we are prototyping this.

However, I don't think we can ship non-determinism in the consensus validator, as it may cause a consensus fork. There may be no way around implementing the JWK voting for now. However, doing so should not be terribly difficult either.

Comment (mystenmark) [reply crates/sui-core/src/consensus_validator.rs:100]: Actually this may be even more tricky - the JWK votes can only be counted in consensus handler, but validator runs before the commit, and hence I don't think can depend on state computed from consensus.

@andll any thoughts?

Comment (andll) [reply crates/sui-core/src/consensus_validator.rs:100]: Ok, I actually found thread about it: 
https://mysten-labs.slack.com/archives/C04FG4Q7YJ3/p1679942005356859

TLDR - the only safe way I can think of if we commit to the signature in the certificate, and therefore do not verify it in the consensus anymore. Now that we are in mainnet this will require some amount of work(thought not impossible)...

File: crates/sui-types/src/crypto.rs [hunk 2]
```diff
@@ -1636,8 +1653,9 @@ pub enum SignatureScheme {
     ED25519,
     Secp256k1,
     Secp256r1,
-    BLS12381, // This is currently not supported for user Sui Address.
     MultiSig,
+    BLS12381, // This is currently not supported for user Sui Address.
```
Comment (mystenmark) [inline crates/sui-types/src/crypto.rs:1657]: we cannot reorder previously existing variants, this will break deserialization of previously committed transactions

Comment (mystenmark) [reply crates/sui-types/src/crypto.rs:1657]: or wait, does this do custom serialization using the flag bytes?

File: crates/sui-types/src/zk_login_authenticator.rs [hunk 3]
```diff
@@ -0,0 +1,232 @@
...
+
+/// Necessary trait for [struct SenderSignedData].
+impl PartialEq for ZkLoginAuthenticator {
+    fn eq(&self, other: &Self) -> bool {
+        self.as_ref() == other.as_ref()
+    }
+}
+
+/// Necessary trait for [struct SenderSignedData].
+impl Eq for ZkLoginAuthenticator {}
```
Comment (benr-ml) [inline crates/sui-types/src/zk_login_authenticator.rs:77]: should we ignore bytes (if set or not), or only compare set bytes?

File: crates/sui-types/src/crypto.rs (hunk 2 above)
Comment (joyqvq) [reply crates/sui-types/src/crypto.rs:1657]: the flag byte is defined below unrelated to the enum serde itself. revised to not change order

File: crates/sui-types/src/crypto.rs [hunk 4]
```diff
@@ -927,19 +927,11 @@ pub trait SuiSignatureInner: Sized + ToFromBytes + PartialEq + Eq + Hash {
...
+    fn get_verification_inputs(&self) -> SuiResult<(Self::Sig, Self::PubKey)> {
+        let pk = Self::PubKey::from_bytes(self.public_key_bytes())
             .map_err(|_| SuiError::KeyConversionError("Invalid public key".to_string()))?;
 
-        let received_addr = SuiAddress::from(&pk);
-        if received_addr != author {
-            return Err(SuiError::IncorrectSigner {
-                error: format!("Signature get_verification_inputs() failure. Author is {author}, received address is {received_addr}")
-            });
-        }
```
Comment (joyqvq) [inline crates/sui-types/src/crypto.rs:941]: this check is moved to verify_generic_secure instead bc sui address is no longer assumed to be derived from pubkey. for zklogin it can be derived from address seed and iss and sub claim name

File: crates/sui-types/src/zk_login_authenticator.rs (hunk 3 above)
Comment (joyqvq) [reply crates/sui-types/src/zk_login_authenticator.rs:77]: yes this is just comparing hte bytes (i believe eq is derived from partial eq)

File: crates/sui-types/src/zk_login_authenticator.rs [hunk 5]
```diff
@@ -0,0 +1,232 @@
...
+        {
+            return Err(SuiError::InvalidSignature {
+                error: "Client id not in whitelist".to_string(),
+            });
+        }
+
+        let pk = RSAPublicKey::from_raw_components(
+            &Base64UrlUnpadded::decode_vec(&selected.n).map_err(|_| {
+                SuiError::InvalidSignature {
+                    error: "Invalid OAuth provider pubkey n".to_string(),
```
[the remark this answers is not in the document]
Comment (joyqvq) [reply crates/sui-types/src/zk_login_authenticator.rs:156]: currently, the pk pulled from google is stored as bytes cached and unverified. optimization here is to cache the RSAPublicKey directly without deser here every time.

File: crates/sui-types/src/message_envelope.rs [hunk 6]
```diff
@@ -29,23 +29,23 @@ pub fn get_google_jwk_bytes() -> Arc<RwLock<Vec<u8>>> {
...
+                        },
+                        {
+                          "alg": "RS256",
+                          "use": "sig",
+                          "n": "1qrQCTst3RF04aMC9Ye_kGbsE0sftL4FOtB_WrzBDOFdrfVwLfflQuPX5kJ-0iYv9r2mjD5YIDy8b-iJKwevb69ISeoOrmL3tj6MStJesbbRRLVyFIm_6L7alHhZVyqHQtMKX7IaNndrfebnLReGntuNk76XCFxBBnRaIzAWnzr3WN4UPBt84A0KF74pei17dlqHZJ2HB2CsYbE9Ort8m7Vf6hwxYzFtCvMCnZil0fCtk2OQ73l6egcvYO65DkAJibFsC9xAgZaF-9GYRlSjMPd0SMQ8yU9i3W7beT00Xw6C0FYA9JAYaGaOvbT87l_6ZkAksOMuvIPD_jNVfTCPLQ==",
+                          "e": "AQAB",
+                          "kty": "RSA",
+                          "kid": "6083dd5981673f661fde9dae646b6f0380a0145c"
+                        }
+                      ]
```
Comment (joyqvq) [inline crates/sui-types/src/message_envelope.rs:48]: pinning a latest JWK

**Review by mskd12 — APPROVED** [review]
LGTM
